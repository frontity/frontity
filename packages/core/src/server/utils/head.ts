import { HelmetData } from "react-helmet";
import { HeadTags } from "../../../types";

export default (helmet: HelmetData): HeadTags => {
  return {
    base: helmet.base.toString(),
    bodyAttributes: helmet.bodyAttributes.toString(),
    htmlAttributes: helmet.htmlAttributes.toString(),
    link: helmet.link.toString(),
    meta: helmet.meta.toString(),
    noscript: helmet.noscript.toString(),
    script: helmet.script.toString(),
    style: helmet.style.toString(),
    title: helmet.title.toString()
  };
};
