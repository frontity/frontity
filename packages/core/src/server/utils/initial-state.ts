import deepmerge from "deepmerge";
import { NormalizedSettings } from "@frontity/file-settings";
import { Package } from "@frontity/types";

export default ({
  settings,
  url
}: {
  settings: NormalizedSettings;
  url: URL;
}) => {
  let state: Package["state"] = {
    frontity: {
      name: settings.name,
      mode: settings.mode,
      platform: "server",
      initial: {
        path: (/^(.*)page\/\d+\/?$/.exec(url.pathname) || [
          null,
          url.pathname
        ])[1],
        page: parseInt(
          (/^.*page\/(\d+)\/?$/.exec(url.pathname) || [null, "1"])[1],
          10
        )
      },
      packages: settings.packages.map(pkg => pkg.name)
    }
  };

  // Merge the initial state with the general state of settings.
  state = deepmerge(settings.state, state, {
    clone: false
  });

  // Merge the state with each package state, in order.
  settings.packages.forEach(pkg => {
    state = deepmerge(state, pkg.state, {
      clone: false
    });
  });

  return state;
};
