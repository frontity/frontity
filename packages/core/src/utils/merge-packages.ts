import { Package } from "@frontity/types";
import deepmerge from "deepmerge";
import getVariable from "./get-variable";

/**
 * A package function.
 *
 * @returns The package settings.
 */
type PackageFunction = () => Package;

/**
 *
 * Overwrites the merge config for deep merge.
 *
 * @param _ - The target array which is ignored.
 * @param sourceArray - The Array to be returned.
 *
 * @returns The source array.
 */
const overwriteMerge: deepmerge.Options["arrayMerge"] = (_, sourceArray) => {
  return sourceArray;
};

/**
 * Type for parameters of the {@link mergePackages} function.
 */
interface MergePackages {
  /**
   * The packages interface to be used.
   */
  packages: {
    [name: string]: Package | PackageFunction;
  };

  /**
   * The state entry.
   */
  state: Package["state"];

  /**
   * The flag for overwritting the arrays.
   */
  overwriteArrays?: boolean;
}

/**
 * Merge all packages together in a single config that can be passed
 * to createStore.
 *
 * @param params - Defined in {@link MergePackages}.
 *
 * @returns A Frontity package that can be passed to the {@link createStore} function.
 */
const mergePackages = ({
  packages,
  state,
  overwriteArrays = false,
}: MergePackages): Package => {
  let config: Package = {
    roots: {},
    state: {},
    actions: {},
    libraries: {
      frontity: {},
    },
    server: {},
  };
  state.frontity.packages.forEach((name) => {
    const variable = getVariable(name, state.frontity.mode);
    const module = packages[variable];

    // If module exist extend the base config.
    if (module) {
      const pkg = typeof module === "function" ? module() : module;
      config = deepmerge(config, pkg, {
        clone: false,
      });
    }
  });
  // Save debug, which is the only value we want to retain.
  const debug = !!config.state?.frontity?.debug;
  // Merge the state from the packages with the initial state.
  config.state = deepmerge(config.state, state, {
    clone: true,
    arrayMerge: overwriteArrays ? overwriteMerge : undefined,
  });
  // Delete the name.
  delete config.name;
  // Restore the debug flag if we are not in production.
  if (process.env.NODE_ENV !== "production")
    config.state.frontity.debug = debug;

  return config;
};

export default mergePackages;
