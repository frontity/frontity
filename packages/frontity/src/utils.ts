import chalk from "chalk";
import { readdir as readDir } from "fs-extra";

export const isPackageNameValid = (name: string): boolean => {
  if (name.length >= 214) return false;
  if ("._".includes(name[0])) return false;
  const specialCharacters = /[/:*?"<>|]/;
  if (specialCharacters.test(name)) return false;
  const nameConventionMatch = /^(?:@[a-z0-9_.-]+\/)?[a-z0-9_.-]+$/;
  return nameConventionMatch.test(name);
};

export const isThemeNameValid = (name: string): boolean => {
  const nameConventionMatch = /^(?:@[\w-]+\/)?[\w-]+$/;
  return nameConventionMatch.test(name);
};

export const errorLogger = (error: Error, message?: string) => {
  console.error(
    chalk.bold.red("\nError: ") +
      chalk.red(error.message) +
      "\n\n" +
      (message || "") +
      `If you need help please visit ${chalk.underline.magenta(
        "https://community.frontity.org/"
      )}.\n`
  );
  process.exit(1);
};

export const isFrontityProjectRoot = async (path: string) => {
  const dirContent = await readDir(path);
  return dirContent.some(content =>
    /^frontity\.settings\.(js|ts)$/i.test(content)
  );
};
