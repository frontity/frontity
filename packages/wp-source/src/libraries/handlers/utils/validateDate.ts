import { ServerError } from "@frontity/source";

/**
 *
 * Check if a given year is a leap year.
 *
 * @param year - Year as integer.
 *
 * @returns If the year is a leap year.
 */
function isLeapYear(year: number) {
  return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
}

/**
 * Check that the given combination of year, month & day is valid.
 *
 * @param year - Year as integer.
 * @param month - Month of the year as integer (should start with 1).
 * @param day - Day of the month as integer (should start with 1).
 *
 */
function validateDate(year: number, month: number, day: number) {
  // This is just an arbitrary sanity check.
  // We ll generously assume that frontity will continue to exist for a hundred years ðŸ˜…
  if (year < 1900 || year > 2100) {
    throw new ServerError(`${year} is a wrong year number`, 404);
  }

  const monthDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (isLeapYear(year)) {
    monthDays[1] = 29;
  }

  if (typeof month === "number" && (month < 1 || month > 12)) {
    throw new ServerError(`${month} is a wrong month number`, 404);
  }
  if (typeof day === "number" && (day < 1 || day > monthDays[month - 1])) {
    throw new ServerError(`${day} is a wrong day number`, 404);
  }
}

export default validateDate;
