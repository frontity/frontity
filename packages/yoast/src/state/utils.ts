import { YoastSocialDefaults, YoastMeta } from "../../types";
import { Author, Taxonomy, PostType, Attachment } from "@frontity/source/";

export type Entity = (Author | Taxonomy | PostType | Attachment) & {
  yoast_meta: YoastMeta;
};

export const getEntity = ({ state, data }): Entity => {
  if (data.isPostType) return state.source[data.type][data.id];
  if (data.isTaxonomy) return state.source[data.taxonomy][data.id];
  if (data.isAuthor) return state.source.author[data.id];
  return null;
};

export const appendPage = ({
  title,
  page,
  totalPages
}: {
  title: string;
  page: number;
  totalPages: number;
}) => (page > 1 ? `${title} - Page ${page} of ${totalPages}` : title);

export const getSocialDefaults = ({ state }): YoastSocialDefaults => {
  const data = state.source.get(state.router.link);

  let type: string;
  let id: number;

  if (data.isArchive) ({ type, id } = data.items[0]);
  if (data.isPostType) ({ type, id } = data);

  if (type && id)
    return state.source[type][id].yoast_meta.yoast_wpseo_social_defaults;

  return {};
};

export const getPathname = route => {
  const [, pathname] = /^(?:(?:[^:/?#]+):)?(?:\/\/(?:[^/?#]*))?([^?#]*)/.exec(
    route
  );
  return pathname;
};

export const getPage = route => {
  const [, page] = /^(?:.*)page\/(\d+)\/?(\?.*)?$/.exec(getPathname(route)) || [
    null,
    "1"
  ];
  return parseInt(page, 10);
};

export const hasName = (
  entity: Entity
): entity is (Author | Taxonomy) & {
  yoast_meta: YoastMeta;
} => "name" in entity;
