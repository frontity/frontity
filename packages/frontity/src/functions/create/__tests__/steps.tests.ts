import {
  normalizeOptions,
  ensureProjectDir,
  createPackageJson,
  createFrontitySettings,
  cloneStarterTheme,
  installDependencies,
  downloadFavicon,
  revertProgress
} from "../steps";
import * as utils from "../../../utils";
import * as fsExtra from "fs-extra";
import * as fetch from "node-fetch";
import * as path from "path";
import * as childProcess from "child_process";
import * as tar from "tar";

jest.mock("../../../utils");
jest.mock("fs-extra");
jest.mock("path");
jest.mock("node-fetch");
jest.mock("child_process");
jest.mock("tar");

const mockedUtils = utils as jest.Mocked<typeof utils>;
const mockedFsExtra = fsExtra as jest.Mocked<typeof fsExtra>;
const mockedFetch = fetch as jest.Mocked<typeof fetch>;
const mockedPath = path as jest.Mocked<typeof path>;
const mockedChildProcess = childProcess as jest.Mocked<typeof childProcess>;
const mockedTar = tar as jest.Mocked<typeof tar>;

beforeEach(() => {
  mockedPath.resolve.mockImplementation((...dirs) => dirs.join("/").replace("./", ""));
})

describe("normalizeOptions", () => {
  beforeEach(() => {
    mockedUtils.isPackageNameValid.mockReset();
  });

  test("returns merged options", () => {
    const defaultOptions = {
      path: "/default/path/to/project",
      typescript: false,
      packages: ["frontity", "@frontity/file-settings"],
      theme: "@frontity/mars-theme"
    };
    const passedOptions = {
      name: "Some Random Name",
      path: "/path/to/project",
      theme: "@frontity/saturn-theme"
    };
    const expectedOptions = {
      name: "some-random-name",
      path: passedOptions.path,
      typescript: defaultOptions.typescript,
      packages: defaultOptions.packages,
      theme: passedOptions.theme
    };
    mockedUtils.isPackageNameValid.mockReturnValue(true);
    const result = normalizeOptions(defaultOptions, passedOptions);
    expect(result).toEqual(expectedOptions);
  });

  test("throws if `options.name` is not valid", () => {
    const defaultOptions = {
      path: "/default/path/to/project",
      typescript: false,
      packages: ["frontity", "@frontity/file-settings"],
      theme: "@frontity/mars-theme"
    };
    const passedOptions = {
      name: "Some, Random Name"
    };
    mockedUtils.isPackageNameValid.mockReturnValue(false);
    expect(() => normalizeOptions(defaultOptions, passedOptions)).toThrow(
      "The name of the package is not valid. Please enter a valid one (only letters and dashes)."
    );
  });
});

describe("ensureProjectDir", () => {
  beforeEach(() => {
    mockedFsExtra.ensureDir.mockReset();
    mockedFsExtra.readdir.mockReset();
    mockedFsExtra.pathExists.mockReset();
  });

  test("works when passing a non existent path", async () => {
    const path = "/path/to/project";
    mockedFsExtra.pathExists.mockImplementation(() => Promise.resolve(false));
    const dirExisted = await ensureProjectDir({ path });
    expect(dirExisted).toBe(false);
    expect(mockedFsExtra.pathExists.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.ensureDir.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.readdir).not.toHaveBeenCalled();
  });

  test("works when passing an existent path with an empty repo", async () => {
    const path = "/path/to/project";
    mockedFsExtra.readdir.mockResolvedValue([
      "README.md",
      ".git",
      ".gitignore",
      "LICENSE"
    ]);
    mockedFsExtra.pathExists.mockImplementation(() => Promise.resolve(true));
    const dirExisted = await ensureProjectDir({ path });
    expect(dirExisted).toBe(true);
    expect(mockedFsExtra.pathExists.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.readdir.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.ensureDir).not.toHaveBeenCalled();
  });

  test("throws when passing an non-empty path", async () => {
    const path = "/path/to/project";
    mockedFsExtra.pathExists.mockImplementation(() => Promise.resolve(true));
    mockedFsExtra.readdir.mockResolvedValue(["file-that-should-not-exist"]);
    await expect(ensureProjectDir({ path })).rejects.toThrow(
      "The directory passed to `create` function is not empty"
    );
  });
});

describe("createPackageJson", () => {
  beforeEach(() => {
    mockedFetch.default.mockReset();
    mockedFetch.default.mockResolvedValue({
      json: () => Promise.resolve({ "dist-tags": { latest: "1.0.0" } })
    } as any);
    mockedFsExtra.writeFile.mockReset();
  });

  test('works with a theme like "@frontity/mars-theme"', async () => {
    const options = {
      name: "random-name",
      theme: "@frontity/mars-theme",
      path: "/path/to/project"
    };
    await createPackageJson(options);
    expect(mockedFetch.default.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.writeFile.mock.calls).toMatchSnapshot();
  });

  test('works with a theme like "mars-theme"', async () => {
    const options = {
      name: "random-name",
      theme: "random-theme",
      path: "/path/to/project"
    };
    await createPackageJson(options);
    expect(mockedFetch.default.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.writeFile.mock.calls).toMatchSnapshot();
  });
});

describe("createFrontitySettings", () => {
  beforeEach(() => {
    mockedFsExtra.readFile.mockReset();
    mockedFsExtra.readFile.mockResolvedValueOnce("$settings$" as any);
    mockedFsExtra.writeFile.mockReset();
  });

  test("works when extension is `js`", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project",
      packages: ["frontity", "@frontity/file-settings"]
    };
    const extension = "js";
    await createFrontitySettings(extension, options);
    expect(mockedFsExtra.readFile).toHaveBeenCalled();
    expect(mockedFsExtra.writeFile.mock.calls).toMatchSnapshot();
  });

  test("works when extension is `ts`", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project",
      packages: ["frontity", "@frontity/file-settings"]
    };
    const extension = "ts";
    await createFrontitySettings(extension, options);
    expect(mockedFsExtra.readFile).toHaveBeenCalled();
    expect(mockedFsExtra.writeFile.mock.calls).toMatchSnapshot();
  });
});

describe("cloneStarterTheme", () => {
  beforeEach(() => {
    mockedFsExtra.readFile.mockReset();
    mockedFsExtra.readFile.mockResolvedValueOnce(JSON.stringify({
      dependencies: {
        "@frontity/mars-theme": "./packages/mars-theme"
      }
    }) as any);
    mockedFsExtra.ensureDir.mockReset();
    mockedUtils.isThemeNameValid.mockReset();
    mockedFsExtra.readdir.mockReset();
    mockedFsExtra.readdir.mockResolvedValueOnce(["file.tgz"]);
    mockedFsExtra.remove.mockReset();
    mockedChildProcess.exec.mockReset();
    (mockedChildProcess as any).exec.mockImplementation(
      (_command: string, _options: object, resolve: Function) => {
        resolve();
      }
    );
    mockedTar.extract.mockReset();
  });

  test("works as expected", async () => {
    const options = {
      path: "/path/to/project",
      theme: "@frontity/mars-theme"
    };
    mockedUtils.isThemeNameValid.mockReturnValue(true);
    await cloneStarterTheme(options);
    expect(mockedFsExtra.readFile.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.ensureDir.mock.calls).toMatchSnapshot();
    expect(mockedUtils.isThemeNameValid.mock.calls).toMatchSnapshot();
    expect(mockedChildProcess.exec.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.readdir.mock.calls).toMatchSnapshot();
    expect(mockedTar.extract.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.remove.mock.calls).toMatchSnapshot();
  });

  test("throws if the theme name is not valid", async () => {
    const options = {
      path: "/path/to/project",
      theme: "@frontity/mars-theme"
    };
    mockedUtils.isThemeNameValid.mockReturnValue(false);
    await expect(cloneStarterTheme(options)).rejects.toThrow(
      "The name of the theme is not a valid npm package name."
    );
  });
});

describe("downloadFavicon", () => {
  beforeEach(() => {
    mockedFetch.default.mockReset();
    (mockedFetch as any).default.mockResolvedValue({
      body: {
        pipe: jest.fn()
      }
    });
    mockedFsExtra.createWriteStream.mockReset();
    (mockedFsExtra as any).createWriteStream.mockReturnValue({
      on: jest.fn((_event, callback) => callback())
    });
  });

  test("works as expected", async () => {
    const options = {
      path: "/path/to/project"
    };
    await downloadFavicon(options);
    expect(mockedFetch.default.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.createWriteStream.mock.calls).toMatchSnapshot();
  });
});

describe("installDependencies", () => {
  beforeEach(() => {
    mockedChildProcess.exec.mockReset();
    (mockedChildProcess as any).exec.mockImplementation(
      (_command: string, _options: object, resolve: Function) => {
        resolve();
      }
    );
  });

  test("works as expected", async () => {
    const options = {
      path: "/path/to/project"
    };
    await installDependencies(options);
    expect(mockedChildProcess.exec.mock.calls).toMatchSnapshot();
  });
});

describe("revertProgress", () => {
  beforeEach(() => {
    mockedFsExtra.readdir.mockReset();
    mockedFsExtra.remove.mockReset();
  });

  test("works if the project directory existed", async () => {
    const dirExisted = true;
    const options = {
      path: "/path/to/project"
    };
    mockedFsExtra.readdir.mockResolvedValue([
      "frontity.settings.js",
      "package.json",
      "packages",
      "favicon.ico",
      "node_modules",
      "package-lock.json"
    ]);
    await revertProgress(dirExisted, options);
    expect(mockedFsExtra.readdir.mock.calls).toMatchSnapshot();
    expect(mockedFsExtra.remove.mock.calls).toMatchSnapshot();
  });

  test("works if the project directory didn't exist", async () => {
    const dirExisted = false;
    const options = {
      path: "/path/to/project"
    };
    await revertProgress(dirExisted, options);
    expect(mockedFsExtra.remove.mock.calls).toMatchSnapshot();
  });
});
