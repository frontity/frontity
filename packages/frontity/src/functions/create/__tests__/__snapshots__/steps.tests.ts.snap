// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cloneStarterTheme works as expected 1`] = `
Array [
  Array [
    "/path/to/project/package.json",
    Object {
      "encoding": "utf8",
    },
  ],
]
`;

exports[`cloneStarterTheme works as expected 2`] = `
Array [
  Array [
    "/path/to/project/packages/mars-theme",
  ],
]
`;

exports[`cloneStarterTheme works as expected 3`] = `
Array [
  Array [
    "@frontity/mars-theme",
  ],
]
`;

exports[`cloneStarterTheme works as expected 4`] = `
Array [
  Array [
    "npm pack @frontity/mars-theme",
    Object {
      "cwd": "/path/to/project/packages/mars-theme",
    },
    [Function],
  ],
]
`;

exports[`cloneStarterTheme works as expected 5`] = `
Array [
  Array [
    "/path/to/project/packages/mars-theme",
  ],
]
`;

exports[`cloneStarterTheme works as expected 6`] = `
Array [
  Array [
    Object {
      "cwd": "/path/to/project/packages/mars-theme",
      "file": "/path/to/project/packages/mars-theme/file.tgz",
      "strip": 1,
    },
  ],
]
`;

exports[`cloneStarterTheme works as expected 7`] = `
Array [
  Array [
    "/path/to/project/packages/mars-theme/file.tgz",
  ],
]
`;

exports[`createFrontitySettings works when extension is \`js\` 1`] = `
Array [
  Array [
    "/path/to/project/frontity.settings.js",
    "{
  \\"name\\": \\"random-name\\",
  \\"state\\": {
    \\"frontity\\": {
      \\"url\\": \\"https://test.frontity.io\\",
      \\"title\\": \\"Test Frontity Blog\\",
      \\"description\\": \\"WordPress installation for Frontity development\\"
    }
  },
  \\"packages\\": [
    \\"@frontity/tiny-router\\",
    {
      \\"name\\": \\"@frontity/mars-theme\\",
      \\"state\\": {
        \\"theme\\": {
          \\"menu\\": [
            [
              \\"Home\\",
              \\"/\\"
            ],
            [
              \\"Nature\\",
              \\"/category/nature/\\"
            ],
            [
              \\"Travel\\",
              \\"/category/travel/\\"
            ],
            [
              \\"Japan\\",
              \\"/tag/japan/\\"
            ],
            [
              \\"About Us\\",
              \\"/about-us/\\"
            ]
          ],
          \\"featured\\": {
            \\"showOnList\\": false,
            \\"showOnPost\\": false
          }
        }
      }
    },
    {
      \\"name\\": \\"@frontity/wp-source\\",
      \\"state\\": {
        \\"source\\": {
          \\"apiUrl\\": \\"https://test.frontity.io/wp-json\\"
        }
      }
    }
  ]
}",
  ],
]
`;

exports[`createFrontitySettings works when extension is \`ts\` 1`] = `
Array [
  Array [
    "/path/to/project/frontity.settings.ts",
    "{
  \\"name\\": \\"random-name\\",
  \\"state\\": {
    \\"frontity\\": {
      \\"url\\": \\"https://test.frontity.io\\",
      \\"title\\": \\"Test Frontity Blog\\",
      \\"description\\": \\"WordPress installation for Frontity development\\"
    }
  },
  \\"packages\\": [
    \\"@frontity/tiny-router\\",
    {
      \\"name\\": \\"@frontity/mars-theme\\",
      \\"state\\": {
        \\"theme\\": {
          \\"menu\\": [
            [
              \\"Home\\",
              \\"/\\"
            ],
            [
              \\"Nature\\",
              \\"/category/nature/\\"
            ],
            [
              \\"Travel\\",
              \\"/category/travel/\\"
            ],
            [
              \\"Japan\\",
              \\"/tag/japan/\\"
            ],
            [
              \\"About Us\\",
              \\"/about-us/\\"
            ]
          ],
          \\"featured\\": {
            \\"showOnList\\": false,
            \\"showOnPost\\": false
          }
        }
      }
    },
    {
      \\"name\\": \\"@frontity/wp-source\\",
      \\"state\\": {
        \\"source\\": {
          \\"apiUrl\\": \\"https://test.frontity.io/wp-json\\"
        }
      }
    }
  ]
}",
  ],
]
`;

exports[`createPackageJson works with a theme like "@frontity/mars-theme" 1`] = `
Array [
  Array [
    "https://registry.npmjs.com/frontity",
  ],
  Array [
    "https://registry.npmjs.com/@frontity/core",
  ],
  Array [
    "https://registry.npmjs.com/@frontity/wp-source",
  ],
  Array [
    "https://registry.npmjs.com/@frontity/tiny-router",
  ],
]
`;

exports[`createPackageJson works with a theme like "@frontity/mars-theme" 2`] = `
Array [
  Array [
    "/path/to/project/package.json",
    "{
  \\"name\\": \\"random-name\\",
  \\"version\\": \\"1.0.0\\",
  \\"private\\": true,
  \\"description\\": \\"Frontity project\\",
  \\"keywords\\": [
    \\"frontity\\"
  ],
  \\"scripts\\": {
    \\"dev\\": \\"frontity dev\\",
    \\"build\\": \\"frontity build\\",
    \\"serve\\": \\"frontity serve\\"
  },
  \\"dependencies\\": {
    \\"frontity\\": \\"^1.0.0\\",
    \\"@frontity/core\\": \\"^1.0.0\\",
    \\"@frontity/wp-source\\": \\"^1.0.0\\",
    \\"@frontity/tiny-router\\": \\"^1.0.0\\",
    \\"@frontity/mars-theme\\": \\"./packages/mars-theme\\"
  }
}
",
  ],
]
`;

exports[`createPackageJson works with a theme like "mars-theme" 1`] = `
Array [
  Array [
    "https://registry.npmjs.com/frontity",
  ],
  Array [
    "https://registry.npmjs.com/@frontity/core",
  ],
  Array [
    "https://registry.npmjs.com/@frontity/wp-source",
  ],
  Array [
    "https://registry.npmjs.com/@frontity/tiny-router",
  ],
]
`;

exports[`createPackageJson works with a theme like "mars-theme" 2`] = `
Array [
  Array [
    "/path/to/project/package.json",
    "{
  \\"name\\": \\"random-name\\",
  \\"version\\": \\"1.0.0\\",
  \\"private\\": true,
  \\"description\\": \\"Frontity project\\",
  \\"keywords\\": [
    \\"frontity\\"
  ],
  \\"scripts\\": {
    \\"dev\\": \\"frontity dev\\",
    \\"build\\": \\"frontity build\\",
    \\"serve\\": \\"frontity serve\\"
  },
  \\"dependencies\\": {
    \\"frontity\\": \\"^1.0.0\\",
    \\"@frontity/core\\": \\"^1.0.0\\",
    \\"@frontity/wp-source\\": \\"^1.0.0\\",
    \\"@frontity/tiny-router\\": \\"^1.0.0\\",
    \\"random-theme\\": \\"./packages/random-theme\\"
  }
}
",
  ],
]
`;

exports[`downloadFavicon works as expected 1`] = `
Array [
  Array [
    "https://favicon.frontity.org/",
  ],
]
`;

exports[`downloadFavicon works as expected 2`] = `
Array [
  Array [
    "/path/to/project/favicon.ico",
  ],
]
`;

exports[`ensureProjectDir works when passing a non existent path 1`] = `
Array [
  Array [
    "/path/to/project",
  ],
]
`;

exports[`ensureProjectDir works when passing a non existent path 2`] = `
Array [
  Array [
    "/path/to/project",
  ],
]
`;

exports[`ensureProjectDir works when passing an existent path with an empty repo 1`] = `
Array [
  Array [
    "/path/to/project",
  ],
]
`;

exports[`ensureProjectDir works when passing an existent path with an empty repo 2`] = `
Array [
  Array [
    "/path/to/project",
  ],
]
`;

exports[`installDependencies works as expected 1`] = `
Array [
  Array [
    "npm install",
    Object {
      "cwd": "/path/to/project",
    },
    [Function],
  ],
]
`;

exports[`revertProgress works if the project directory didn't exist 1`] = `
Array [
  Array [
    "/path/to/project",
  ],
]
`;

exports[`revertProgress works if the project directory existed 1`] = `
Array [
  Array [
    "/path/to/project",
  ],
]
`;

exports[`revertProgress works if the project directory existed 2`] = `
Array [
  Array [
    "/path/to/project/frontity.settings.js",
  ],
  Array [
    "/path/to/project/package.json",
  ],
  Array [
    "/path/to/project/packages",
  ],
  Array [
    "/path/to/project/favicon.ico",
  ],
  Array [
    "/path/to/project/node_modules",
  ],
  Array [
    "/path/to/project/package-lock.json",
  ],
]
`;
